docs: add mode and mtime and .touch/.chmod mfs commands (#572)
ipfs/interface-js-ipfs-core: SPEC/FILES.md

{
    path: '/tmp/myfile.txt', // The file path
    content: <data> // A Buffer, Readable Stream, Pull Stream or File with the contents of the file

>     mode: '0755' // optional string or integer mode to store the entry with. strings will be interpreted as a base 8 number

Reviewer: When I think about a unix mode as a string I think “rwx” “r--“ etc. Why not use this format as our stringy mode?

Author: Yeah?  I have to admit I don't when setting modes, only when viewing them with `ls -l` or whatever.

In my head at least, the choice was between a literal base 10 number (e.g. `493`) or a literal base 8 number (e.g. `0755`).  Linters don't seem to like literal numbers in anything other than base 10 so that's why it's a string `'0755'` above.

It's good to be flexible though, maybe something we could add later?

Reviewer: If it helps to get this stuff merged faster then sure 👍 We’ll be supporting rwx format in chmod so I reckon people will expect to be able to use it here too.

Author: We support Symbolic Notation in `chmod` (e.g. `a=rwx`, `ug-x`, etc) which are modifications to an existing `mode`.

I guess if you want to use it here you could apply the modification to the default mode for files/directories?

>     mtime: <time> // A Date object, an object with `{ sec, nsecs }` properties or the output of `process.hrtime()` or `process.hrtime.bigint()`

Reviewer 2: optional?

Author: Fixed!

{
  path: '/tmp/myfile.txt', // The file path
  content: <data> // A Buffer, Readable Stream, Pull Stream or File with the contents of the file
  mode: '0755' // optional string or integer mode to store the entry with. strings will be interpreted as a base 8 number

>   `mtime` Either a ` Date` object, an object with `{ sec, nsecs }` properties or the output of `process.hrtime()`  (default: now)

Reviewer 2: `mtime:`?

Also, isn't this also optional?

Author: Fixed!

>   mtime: <time> // A Date object, an object with `{ sec, nsecs }` properties or the output of `process.hrtime()` or `process.hrtime.bigint()

Reviewer: ```suggestion
  mtime: <time> // A `Date` object or an object with `{ sec, nsecs }` properties or the output of `process.hrtime()` or `process.hrtime.bigint()
```

Author: Same as above re: https://github.com/ipfs/aegir/issues/487

```suggestion
  - `mtime` Either a ` Date` object, an object with `{ sec, nsecs }` properties or the output of `process.hrtime()`  (default: now)
```

Reviewer: +1 with the wording but this needs formatting appropriately for the code block it’s in.

The Mutable File System (MFS) is a virtual file system on top of IPFS that exposes a Unix like API over a virtual directory. It enables users to write and read from paths without having to worry about updating the graph. It enables things like [ipfs-blob-store](https://github.com/ipfs/ipfs-blob-store) to exist.

#### `files.chmod`

> Change mode for files and directories

##### `ipfs.files.chmod(path, mode, [options])`

Where:

- `path` is the path to the entry to modify.  It might be:
  - An existing MFS path to a file or a directory (e.g. `/my-dir/my-file.txt`)
  - An IPFS path (e.g. `/ipfs/QmWGeRAEgtsHW3ec7U4qW2CyVy7eA2mFRVbk1nb24jFyks`)
  - A [CID][cid] instance (e.g. `new CID('QmWGeRAEgtsHW3ec7U4qW2CyVy7eA2mFRVbk1nb24jFyks')`)
- `mode` is the new file mode.  It might be:
  - A string octal, e.g. `'0755'`

>   - A string modification of the existing mode, e.g. `'+x'`, `'-gw'`, etc

Reviewer 2: ```suggestion
  - A string modification of the existing mode, e.g. `'a+x'`, `'g-w'`, etc
```

>   - An integer, e.g. the returned value from `parseInt('0755', 8)`

Reviewer 2: Or just `0o755`? (I'm not sure if supporting the string version is really necessary given that JS has octal support but it may make for nicer JSON)?

Author: Octals-in-strings are supported because the recommended eslint configuration [disallows octal literals](https://eslint.org/docs/rules/no-octal) - our users are likely to be passing them around as strings instead as that's what eslint deems to be the 'correct' way to do it.

Reviewer 2: It disallows octal literals of the form '07...' but I'm pretty sure '0o7...' (ES6) is perfectly fine.

Author: Fair point - we can drop the string octal stuff as args in that case.

At the moment the js implementation uses string octals for mode values returned from the various RPC HTTP endpoints for the sake of readability, happy to keep that?

- `options` is an optional Object that might contain the following keys:
  - `recursive` is a Boolean value that indicates if `mode` should be applied to all sub files/directories of `path` (default: false)
  - `hashAlg` is which algorithm to use when creating CIDs for modified entries. (default: `sha2-256`) [The list of all possible values]( https://github.com/multiformats/js-multihash/blob/master/src/constants.js#L5-L343)
  - `flush` is a Boolean value to decide whether or not to immediately flush MFS changes to disk (default: true)
  - `cidVersion`: the CID version to use for any updated entries (integer, default 0)

**Returns**

| Type | Description |
| -------- | -------- |
| `Promise<void>` | If action is successfully completed. Otherwise an error will be thrown |

**Example:**

```JavaScript
// To give a file -rwxrwxrwx permissions
await ipfs.files.chmod('/path/to/file.txt', parseInt('0777', 8))

// Alternatively

> await ipfs.files.chmod('/path/to/file.txt', '+rwx')

Reviewer 2: `?+rwx`?

Author: Not sure I follow - what does the leading `?` mean in `?+rwx`?  I can't see any reference to it in `man chmod` and when I try to do this locally I get an error:

```console
$ chmod ?+rwx foo.txt
chmod: Invalid file mode: ?+rwx
```

Reviewer 2: Ah, nevermind. I didn't realize `+rwx` was equivalent to `a+rwx`.

Reviewer: That's a good point, it will support `u+rwx`, `g+r`, `o+w`, `a-w` etc. though right? We should specify what subset of https://en.wikipedia.org/wiki/Modes_(Unix)#Format we do support.

Author: Yes, it'll support [all of those](https://github.com/ipfs/js-ipfs-mfs/pull/63/files#diff-e68b618aab9b596a8f2ed03586f2ef2aR79-R151).

Though now you mention it there's a testing gap for `a...`

..and maybe not `X` right now, though it can be added.

The Mutable File System (MFS) is a virtual file system on top of IPFS that exposes a Unix like API over a virtual directory. It enables users to write and read from paths without having to worry about updating the graph. It enables things like [ipfs-blob-store](https://github.com/ipfs/ipfs-blob-store) to exist.

#### `files.chmod`

> Change mode for files and directories

##### `ipfs.files.chmod(path, mode, [options])`

Where:

- `path` is the path to the entry to modify.  It might be:
  - An existing MFS path to a file or a directory (e.g. `/my-dir/my-file.txt`)
  - An IPFS path (e.g. `/ipfs/QmWGeRAEgtsHW3ec7U4qW2CyVy7eA2mFRVbk1nb24jFyks`)
  - A [CID][cid] instance (e.g. `new CID('QmWGeRAEgtsHW3ec7U4qW2CyVy7eA2mFRVbk1nb24jFyks')`)
- `mode` is the new file mode.  It might be:
  - A string octal, e.g. `'0755'`
  - A string modification of the existing mode, e.g. `'+x'`, `'-gw'`, etc
  - An integer, e.g. the returned value from `parseInt('0755', 8)`
- `options` is an optional Object that might contain the following keys:
  - `recursive` is a Boolean value that indicates if `mode` should be applied to all sub files/directories of `path` (default: false)

>   - `format` is what type of nodes to write any modified entries as (default: `dag-pb`)

Reviewer: Is this actually possible? What formats are supported? I’ve seen this in the docs for the other commands but always wondered...

Author: It's possible if you specify `dag-cbor` but the thing you'd get back probably isn't readable as you'd be expecting an instance of `DAGLink` from the `ipld-dag-pb` module and you'd get a plain JS object back.

Like this it's consistent with our other commands but it's a footgun so we should get rid of this everywhere and just make IPFS things be `dag-pb`.

Reviewer: Can you remove from the docs here then and either also remove the others or make an issue so we can track it? 🙏

Author: I've pulled it out of this PR and opened #575 with removals from the other commands because `mode`/`mtime` discussions should not block that.

// }
```

#### `files.touch`

> Update the mtime of a file or directory

##### `ipfs.files.touch(path, [options])`

Where:

- `path` is the path to the file or directory to update. It might be:
  - An existing MFS path to a file or directory (e.g. `/my-dir/a.txt`)
  - An IPFS path (e.g. `/ipfs/QmWGeRAEgtsHW3ec7U4qW2CyVy7eA2mFRVbk1nb24jFyks`)
  - A [CID][cid] instance (e.g. `new CID('QmWGeRAEgtsHW3ec7U4qW2CyVy7eA2mFRVbk1nb24jFyks')`)
- `options` is an optional Object that might contain the following keys:

>   - `mtime` A Date object, an object with `{ sec, nsecs }` properties or the output of `process.hrtime()` or `process.hrtime.bigint()`  (default: now)

Reviewer: ```suggestion
  - `mtime` A ` Date` object or an object with `{ sec, nsecs }` properties or the output of `process.hrtime()` or `process.hrtime.bigint()`  (default: now)
```

Author: TBH we can't use `process.hrtime.bigint()` until https://github.com/ipfs/aegir/issues/487 is resolved.

```suggestion
  - `mtime` Either a ` Date` object, an object with `{ sec, nsecs }` properties or the output of `process.hrtime()`  (default: now)
```
